File: retrieval-service/src/main/java/com/comeback/researchplatform/retrievalservice/ratelimit/TokenBucket.java
package com.comeback.researchplatform.retrievalservice.ratelimit;

import java.util.function.LongSupplier;

/**
 * One bucket, for one domain.
 * <p>
 * Holds at most {@code capacity} tokens (the burst) and regains {@code refillPerSecond}
 * of them per second. Every request spends one. No background timer refills it — instead
 * each call works out how much time has passed since the last look and converts that
 * into tokens, which is the same result for none of the cost.
 */
public class TokenBucket {

    private final double capacity;
    private final double refillPerSecond;
    private final LongSupplier clockMillis;

    private double tokens;
    private long lastRefillMillis;

    public TokenBucket(double capacity, double refillPerSecond, LongSupplier clockMillis) {
        this.capacity = capacity;
        this.refillPerSecond = refillPerSecond;
        this.clockMillis = clockMillis;
        // A domain nobody has touched yet is not owed any wait, so it starts full.
        this.tokens = capacity;
        this.lastRefillMillis = clockMillis.getAsLong();
    }

    /**
     * @return true if a token was available and has been spent; false if the caller
     *         must back off. Never blocks — waiting is the caller's decision.
     */
    public synchronized boolean tryAcquire() {
        refill();
        // Not "> 0": holding 0.3 of a token is not enough to make a whole request.
        if (tokens >= 1.0) {
            tokens -= 1.0;
            return true;
        }
        return false;
    }

    private void refill() {
        long now = clockMillis.getAsLong();
        long elapsedMillis = now - lastRefillMillis;

        // Move the marker first, unconditionally. If the clock jumped backwards (NTP
        // correction), this resyncs immediately; leaving it stale would make elapsed
        // negative on every future call and freeze the bucket until real time caught up.
        lastRefillMillis = now;

        if (elapsedMillis <= 0) {
            return;
        }

        // Millis to seconds, and 1000.0 not 1000 — integer division would floor every
        // sub-second gap to zero and the bucket would never refill under steady traffic.
        double earned = (elapsedMillis / 1000.0) * refillPerSecond;

        // Math.min is the idle cap: eight untouched hours must not bank 28,800 tokens.
        tokens = Math.min(capacity, tokens + earned);
    }
}

File: retrieval-service/src/main/resources/scripts/token_bucket.lua
-- Token bucket, one domain per key. Same algorithm as the old in-memory TokenBucket,
-- moved into Redis so every retrieval-service instance draws from the SAME bucket.
--
-- Redis runs a script start to finish with nothing interleaved, so read-check-write
-- here is atomic. That is the whole reason this is Lua and not three RestTemplate calls.
--
-- KEYS[1] = bucket key, e.g. "ratelimit:v1:thehindu.com"
-- ARGV[1] = capacity (burst)   ARGV[2] = refill tokens per second
-- returns 1 if a token was spent, 0 if the caller must back off

local key      = KEYS[1]
local capacity = tonumber(ARGV[1])
local rate     = tonumber(ARGV[2])

-- Time comes from Redis, not from the caller. Every instance shares this one clock,
-- so a JVM whose clock drifts a second cannot corrupt the shared bucket.
local t   = redis.call('TIME')
local now = tonumber(t[1]) * 1000 + tonumber(t[2]) / 1000

local data         = redis.call('HMGET', key, 'tokens', 'ts')
local tokens       = tonumber(data[1])
local lastRefillMs = tonumber(data[2])

-- Nothing stored: a domain nobody has touched is not owed any wait, so it starts full.
if tokens == nil then
  tokens = capacity
  lastRefillMs = now
end

-- Refill. Elapsed is clamped at 0 so a backwards clock jump cannot drain the bucket.
local elapsedMs = math.max(0, now - lastRefillMs)
tokens = math.min(capacity, tokens + (elapsedMs / 1000.0) * rate)

-- Not "> 0": 0.3 of a token is not enough to make a whole request.
local allowed = 0
if tokens >= 1.0 then
  tokens = tokens - 1.0
  allowed = 1
end

redis.call('HSET', key, 'tokens', tokens, 'ts', now)

-- Once capacity/rate seconds pass, an untouched bucket has refilled to full — which is
-- exactly what a missing key gives us. So past that point the key carries no information
-- and can go. ceil, and never less than 1, since EXPIRE takes whole seconds.
redis.call('EXPIRE', key, math.max(1, math.ceil(capacity / rate)))

return allowed
